File: Baekjoon/Java/외계인의 기타 연주.java
import java.io.*;
import java.util.*;

public class Main {
    static int ans;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int P = Integer.parseInt(st.nextToken());

        List<List<Integer>> strings = new ArrayList<>();
        ans = 0;
        for (int i = 0; i < 7; i++) {
            strings.add(new ArrayList<Integer>());
        }

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int num = Integer.parseInt(st.nextToken());
            int fret = Integer.parseInt(st.nextToken());
            List<Integer> string = strings.get(num);
            if (string.isEmpty()) {
                string.add(fret);
                ans++;
            } else {
                if (fret == string.get(string.size() - 1)) {
                    string.remove(string.size() - 1);
                    ans++;
                } else if (fret > string.get(string.size() - 1)) {
                    string.add(fret);
                    ans++;
                } else {
                    while (!string.isEmpty() && string.get(string.size() - 1) > fret) {
                        string.remove(string.size() - 1);
                        ans++;
                    }
                    if (!string.isEmpty() && fret == string.get(string.size() - 1)) {
                        continue;
                    }
                    string.add(fret);
                    ans++;
                }
            }
        }
        System.out.println(ans);
    }
}

File: Baekjoon/Python/외계인의 기타 연주.py
import sys
input = sys.stdin.readline

N, P = map(int,input().split())
melody = [list(map(int,input().split())) for _ in range(N)]

strings = [[] for _ in range(7)]
ans = 0

for l, p in melody:
    if len(strings[l]) == 0:
        strings[l].append(p)
        ans += 1
    else:
        if p == strings[l][-1]:
            continue
        elif p > strings[l][-1]:
            strings[l].append(p)
            ans += 1
        else:
            while strings[l] and strings[l][-1] > p:
                strings[l].pop()
                ans += 1
            if strings[l] and strings[l][-1] == p:
                continue
            strings[l].append(p)
            ans += 1

print(ans)